teleport: 11.3.4 -> 12.0.2, reintroduce teleport_11

justinas opened the pull request:
###### Description of changes

New major version.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

justinas committed:
teleport: 11.3.4 -> 12.0.2

justinas commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -105,7 +105,7 @@ In addition to numerous new and upgraded packages, this release has the followin
 - The EC2 image module previously detected and automatically mounted ext3-formatted instance store devices and partitions in stage-1 (initramfs), storing `/tmp` on the first discovered device. This behaviour, which only catered to very specific use cases and could not be disabled, has been removed. Users relying on this should provide their own implementation, and probably use ext4 and perform th ...
 
-- `teleport` has been upgraded to major version 11. Please see upstream [upgrade instructions](https://goteleport.com/docs/setup/operations/upgrading/) and [release notes](https://goteleport.com/docs/changelog/#1100).
+- `teleport` has been upgraded from major version 10 to major version 12. Please see upstream [upgrade instructions](https://goteleport.com/docs/setup/operations/upgrading/) and release notes for versions [11](https://goteleport.com/docs/changelog/#1100) and [12](https://goteleport.com/docs/changelog/#1201). Note that Teleport does not officially support upgrades across more than one major versio ...
```
This was going to catch up with us eventually: Teleport has a more frequent release schedule (major version every ~3 months versus NixOS 6 months). 

I am not sure whether this warrants `stateVersion` and carrying several package versions (`teleport_10`, `teleport_11`, `teleport_12`). Teleport does not require manual steps in Postgres, but *in theory* something could break when upgrading 10 -> 12 (hence the recommendation to use 11 before upgrading to 12).

One more question: is it customary to link to a PR/commit in the release notes? It would be nice to link directly to https://github.com/NixOS/nixpkgs/pull/214425 here.

arianvp replied:
exposing `teleport_11` would be nice.

arianvp replied:
Linking to PR/commits in release notes is fine! Adds more context!

justinas replied:
> exposing `teleport_11` would be nice.

Will try to add multiple versions. Couple of questions:
* Do we add back `teleport_10` as well? Teleport supports latest + 2 previous major versions (10, 11, 12 currently).
* Do we update `nixosTests.teleport` to test each major version?

justinas committed:
teleport_11: init at 11.3.5

justinas committed:
nixos/teleport: add "package" option

justinas commented on nixos/modules/services/networking/teleport.nix:
```diff
@@ -74,14 +82,14 @@ in
 
   config = mkIf config.services.teleport.enable {
-    environment.systemPackages = [ pkgs.teleport ];
+    environment.systemPackages = [ cfg.package ];
```
@ymatsiuk, you originally introduced the module: is this line required at all? If the service needs `teleport` in PATH (e.g. to re-exec itself), we could add it to the systemd unit path. Otherwise, I think it is conventional to not implicitly mess with `environment.systemPackages` if not necessary.

ymatsiuk replied:
>Otherwise, I think it is conventional to not implicitly mess with `environment.systemPackages` if not necessary.

I can't agree here, even though I have no strong opinion on this, it feels natural to have tools provided by the package when you enable its service, especially if the package provides more than one binary necessary to interact with the running service.

justinas commented:
Please see the updated PR.

* Added `teleport_11` alongside the updated default `teleport` (12)
* Added `package` option to the module
* Updated release notes to recommend settings this option to upgrade to v11.

ymatsiuk commented on nixos/tests/teleport.nix:
```diff
@@ -64,36 +75,41 @@ let
+    "minimal_${name}" = makeTest {
+      # minimal setup should always work
+      name = "teleport-minimal-setup";
+      meta.maintainers = with pkgs.lib.maintainers; [ ymatsiuk ];
```
```suggestion
      meta.maintainers = with pkgs.lib.maintainers; [ justinas ];
```

Do you mind? I'm not using teleport anymore.

justinas committed:
nixos/tests/teleport: make justinas the maintainer

justinas commented:
* Updated `teleport_11` from 11.3.4 to 11.3.5
* Replaced ymatsiuk as the maintainer for teleport tests, per request.

ymatsiuk reviewed (approved):
Thanks for your effort! 👍🏻

takeda commented:
@justinas this package fails to compile on darwin, it looks like the issue mentioned here: https://github.com/NixOS/nixpkgs/pull/214425#issuecomment-1416721340

Here's my build output:

```yarn run v1.22.19
$ yarn workspace @gravitational/teleport build --output-path=../../../webassets/teleport/app
$ g-build --config webpack.prod.config.js --progress --bail --output-path=../../../webassets/teleport/app
assets by status 3.92 KiB [cached] 4 assets2mginugin
Entrypoint app = runtime~app.d96ed60b58b56df22960.js app.33babda0435fe6278445.js
runtime modules 2.5 KiB 3 modules
./src/boot.tsx 1.22 KiB [built] [code generated] [1 error]

ERROR in ./src/boot.tsx 41:2
Module parse failed: Unexpected token (41:2)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|
| ReactDOM.render(
>   <Teleport history={history.original()} ctx={teleportContext} />,
|   document.getElementById('app')
| );

webpack 5.73.0 compiled with 1 error in 2552 ms
✨  Done in 26.90s.
buildPhase completed in 9 minutes 18 seconds
installing
cp: cannot stat 'webassets/.': No such file or directory
error: builder for '/nix/store/q1pp1875j0gcjcgsri45ksplg5ll8srx-teleport-webassets-12.0.2.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/qhiav5bc3crjra9b8a0vvnlnsi09d3mn-teleport-12.0.2.drv' failed to build
error: program '/nix/store/45ws0jiwkilqcg6mwnpp5rz5danvm0ll-nix-2.11.0/bin/nix-shell' failed with exit code 100
```

justinas commented:
Hi @takeda ,

I'm terribly sorry about this! The upstream fix for the web assets made it into 11.3.4, but not 12.0.2 (which we packaged up here) - it was only made available in 12.0.3 onwards.

In any case, I've opened a PR to fix: https://github.com/NixOS/nixpkgs/pull/219971 .
